db/repair.go

```diff
@@ -0,0 +1,264 @@
+//   CloudController, but no events exist, then a START event is created for T0
+// * If a service_instance_guid is currently in a CREATED state according to
+//   CloudController, but no events exist, then a CREATED event is created for T0
+func (pc *PostgresClient) RepairEvents(cfClient cf.Client) (err error) {
+	tx, txErr := pc.Conn.Begin()
+	if txErr != nil {
+		return txErr
+	}
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+			return
+		}
+		err = tx.Commit()
+	}()
+
+	// remove any events with id=0
+	// FIXME: maybe move to migration if done at all?
+	_, err = tx.Exec(`delete from app_usage_events where id = 0`)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(`delete from service_usage_events where id = 0`)
+	if err != nil {
+		return err
+	}
+
+	// fetch the most recent event timestamp
+	var firstEventTimestamp *time.Time
+	err = tx.QueryRow(`
+		with
+		event_dates as (
+			select created_at from app_usage_events
+			union
+			select created_at from service_usage_events
+		)
+		select
+			min(created_at::timestamptz)
+		from
+			event_dates
+		where
+			created_at is not null
+	`).Scan(&firstEventTimestamp)
+	if err != nil {
+		return err
+	}
+	if firstEventTimestamp == nil {
+		return errors.New("Database appears to be empty and thus cannot be repaired.")
+	}
+
+	// fetch list of app guids we have events for
+	rows, err := tx.Query(`
+		select distinct
+			raw_message->>'app_guid'
+		from
+			app_usage_events
+		where
+			raw_message->>'app_guid' is not null
+	union
+		select distinct
+			raw_message->>'service_instance_guid'
+		from
+			service_usage_events
+		where
+			raw_message->>'service_instance_guid' is not null
+	`)
+	if err != nil {
+		return err
+	}
+	knownGuids := map[string]bool{}
+	for rows.Next() {
+		var guid string
+		if err := rows.Scan(&guid); err != nil {
+			return err
+		}
+		knownGuids[guid] = true
+	}
+	if err := rows.Err(); err != nil {
+		return err
+	}
+
+	// cache all the things
+	orgs, err := cfClient.GetOrgs()
+	if err != nil {
+		return err
+	}
+	spaces, err := cfClient.GetSpaces()
+	if err != nil {
+		return err
+	}
+	services, err := cfClient.GetServices()
+	if err != nil {
+		return err
+	}
+	servicePlans, err := cfClient.GetServicePlans()
+	if err != nil {
+		return err
+	}
+
+	// fetch all the running apps
+	apps, err := cfClient.GetApps()
+	if err != nil {
+		return err
+	}
+	for _, app := range apps {
+		// Skip non running apps (FIXME: ListApps() could probably filter this)
+		if app.State != StateStarted {
+			continue
+		}
+		// Skip anything updated since we started collecting events
+		updated, err := time.Parse(time.RFC3339, app.UpdatedAt)
+		if err != nil {
+			return err
+		}
+		fmt.Println("app", app.Guid)
+		// Skip if we already have events for this app
+		if seen := knownGuids[app.Guid]; seen {
+			continue
+		}
+		fmt.Println("detected missing app", app.Guid)
+		if updated.After(*firstEventTimestamp) {
```

Comment:
yes we _should_ have an event, so we skip this app since if for some reason we got to this point then starting creating fake events it would be incorrect.

I'm not sure exactly when that case _could_ occur... I only know that it should be skipped if it _did_ occur :)

One possible situation is if an app was updated but the event hasn't arrived yet.